### ISSUE TITLE
[Bug]: How to use pdb with RLinf?

### ISSUE THREAD

**Issue by JYounngS** [RLinf/RLinf#383]
### Affected component

I inserted a pdb.set_trace() breakpoint in my code, but when the program hits the breakpoint, the terminal appears to hang—it doesn’t print the debugger prompt or allow me to inspect variables.

### Bug description

I'm trying to debug my training script using Python’s built-in debugger (pdb). I set 0-0 to restrict execution to a single GPU and launched training in the ManiSkill simulation environment with the following command:

`bash examples/embodiment/run_embodiment.sh maniskill_ppo_openvlaoft_quickstart`

I inserted a pdb.set_trace() breakpoint in my code, but when the program hits the breakpoint, the terminal appears to hang—it doesn’t print the debugger prompt or allow me to inspect variables.

<img width="652" height="128" alt="Image" src="https://github.com/user-attachments/assets/e58a8506-2643-43eb-a14d-8a8f232796bd" />

Is there a recommended way to use pdb (or an alternative debugger) with RLinf, especially when running through shell scripts like this?

Looking forward to your reply!

### Configuration YAML file

defaults:
  - env/train: PutOnPlateInScene25Main
  - env/eval: maniskill_ood_template
  - override hydra/job_logging: stdout

hydra:
  run:
    dir: .
  output_subdir: null 
  searchpath:
    - file://${oc.env:EMBODIED_PATH}/config/

cluster:
  num_nodes: 1
  component_placement:
    actor,env,rollout: 0-3

runner:
  task_type: embodied
  logger:
    log_path: "../results"
    project_name: rlinf
    experiment_name: "test_openvla"
    logger_backends: ["tensorboard"] # wandb, swanlab

  max_epochs: 1000
  max_steps: -1

  only_eval: False
  val_check_interval: -1
  save_interval: 40
  seq_length: 4096
  max_prompt_length: 30

algorithm:
  auto_reset: True
  ignore_terminations: False
  use_fixed_reset_state_ids: False
  normalize_advantages: True
  kl_penalty: kl  # how to estimate kl divergence: kl or kl_penalty
  group_size: 1
  n_chunk_steps: 20
  n_eval_chunk_steps: 10
  # training rollout mbs
  rollout_micro_batch_size: 64
  num_group_envs: 8
  rollout_epoch: 1

  reward_type: action_level
  logprob_type: token_level
  entropy_type: token_level

  # mbs to do log prob inference, can be set to
  # lower than rollout_micro_batch_size to reduce
  # memory usage
  logprob_forward_micro_batch_size: 8 # ${.rollout_micro_batch_size}

  adv_type: gae
  loss_type: actor_critic
  loss_agg_func: "token-mean"
  kl_beta: 0.0
  entropy_bonus: 0
  clip_ratio_high: 0.28
  clip_ratio_low: 0.2
  clip_ratio_c: 3.0
  value_clip: 0.2
  
  huber_delta: 10.0

  gamma: 0.99
  gae_lambda: 0.95

  # params for rollout
  sampling_params:
    use_greedy: False
    temperature_train: 1.0
    temperature_eval: 0.6
    top_k: 0
    top_p: 1.0
    repetition_penalty: 1.0

  # length argument for autoregressive sampling
  # max length means max amount of tokens to generate
  length_params:
    max_new_token: null
    max_length: 1024
    min_length: 1

env:
  group_name: "EnvGroup"
  channel:
    name: "env_buffer_list"
    queue_name: "obs_buffer"
    queue_size: 0
  enable_offload: False

rollout:
  group_name: "RolloutGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "action_buffer"
    queue_size: 0
  mode: "colocate"
  backend: "huggingface"
  model_dir: "/workspace/RLinf/ckpts/Openvla-oft-SFT-libero10-trajall/"
  precision: ${actor.model.precision}
  gpu_memory_utilization: 0.5  # not used in Embodied
  enforce_eager: True  # not used in Embodied
  enable_offload: True
  pipeline_stage_num: 1

actor:
  group_name: "ActorGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "replay_buffer"
    queue_size: 0
  training_backend: "fsdp"
  checkpoint_load_path: "/workspace/RLinf/ckpts/Openvla-oft-SFT-libero10-trajall/"
  checkpoint_save_path: "../results"
  micro_batch_size: 10
  global_batch_size: 80
  seed: 1234
  enable_offload: True

  tokenizer:
    tokenizer_type: "HuggingFaceTokenizer"
    tokenizer_model: "/workspace/RLinf/ckpts/Openvla-oft-SFT-libero10-trajall/"
    extra_vocab_size: 421
    use_fast: False
    trust_remote_code: True
    padding_side: "right"

  model:
    model_name: "openvla_oft"
    value_type: ${algorithm.reward_type}  # 'action' or 'token'
    action_dim: 7
    num_action_chunks: 8
    use_proprio: False
    unnorm_key: bridge_orig
    val_micro_batch_size: 8
    center_crop: True
    do_sample: True
    
    precision: "bf16"
    add_bias_linear: False
    add_qkv_bias: True
    vocab_size: 32000
    hidden_size: 4096
    policy_setup: "widowx_bridge"
    add_value_head: True
    image_size: [224, 224]
    is_lora: True
    lora_rank: 32
    lora_path: /workspace/RLinf/ckpts/oft-sft/lora_004000
    ckpt_path: null
    use_wrist_image: False
    attn_implementation: "flash_attention_2"
    low_cpu_mem_usage: True
    trust_remote_code: True
    gradient_checkpointing: False

  optim:
    lr: 1.0e-4
    value_lr: 3.0e-3
    adam_beta1: 0.9
    adam_beta2: 0.999
    adam_eps: 1.0e-05
    clip_grad: 10.0
    critic_warmup_steps: 0

  fsdp_config:                  # below are fsdp configs, for more details, see https://pytorch.org/docs/stable/fsdp.html and https://huggingface.co/docs/accelerate/package_reference/fsdp

    strategy: "fsdp"            # FSDP strategy: ["fsdp", "fsdp2"]

    # Sharding strategy: "full_shard" (shard parameters, gradients, optimizer states),
    # "shard_grad_op" (shard gradients and optimizer states only), 
    # "hybrid_shard" (combines data parallelism and model parallelism - sharding within FSDP groups, replicate across DDP groups),
    # "no_shard" (no sharding)
    sharding_strategy: "full_shard" 

    cpu_offload: False           # whether to offload parameters and gradients to CPU when not in use，if True, actor's enable_offload should be True too.
    offload_pin_memory: False    # whether FSDP2's CPU offload policy should pin memory when cpu_offload is True
    reshard_after_forward: True  # if True, FSDP2 will reshard parameters after forward pass to save memory

    enable_gradient_accumulation: True  # if True, gradient accumulation will be enabled for FSDP/FSDP2 training, which will enhance training performance but increase memory usage.
    forward_prefetch: False             # if True, FSDP will explicitly prefetches the next upcoming all-gather while executing in the forward pass. only use with static graphs. Overlaps communication with computation to improve performance.
    limit_all_gathers: False            # if True, FSDP will synchronizes CPU threads to limit the number of concurrent all-gathers. Only affects strategies that schedule all-gathers
    backward_prefetch: null             # options are null, 'pre', 'post'. if 'pre', FSDP will prefetch the next upcoming all-gather while computing gradients. if 'post', FSDP will prefetch the next all-gather until current gradient is computed.
    use_orig_params: False              # if True, FSDP will use module's original parameters, it means it will expose nn.Module.named_parameters rather than FlatParameter
    use_liger_kernel: False             # if True, liger_kernel will be used for FSDP, note that currently supported models in RLinf are [qwen2.5, qwen2.5-vl], for more details, see liger_kernel's doc

    fsdp_size: -1               # Number of GPUs per FSDP group for hybrid sharding. -1 means use all available GPUs in a single FSDP group

    mixed_precision:            # mixed precision settings for fsdp/fsdp2
      param_dtype: ${actor.model.precision}
      reduce_dtype: ${actor.model.precision}
      buffer_dtype: ${actor.model.precision}

    amp:
      enabled: False                 # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
      precision: "bf16"              # precision for AMP, options are ["fp16" , "bf16"]
      use_grad_scaler: False         # if True, GradScaler will be used for AMP training to prevent underflow.

reward:
  use_reward_model: False

critic:
  use_critic_model: False

### Log file

The program interruption was caused by the user themselves.

### Environment

python -V:
Python 3.11.14
uv pip list:
Using Python 3.11.14 environment at: /opt/venv/openvla-oft
Package                            Version
---------------------------------- -----------
absl-py                            2.3.1
accelerate                         1.11.0
aiohappyeyeballs                   2.6.1
aiohttp                            3.13.2
aiohttp-cors                       0.8.1
aiosignal                          1.4.0
annotated-doc                      0.0.3
annotated-types                    0.7.0
antlr4-python3-runtime             4.11.1
anyio                              4.11.0
argcomplete                        3.6.3
arm-pytorch-utilities              0.4.3
array-record                       0.8.2
asttokens                          3.0.0
astunparse                         1.6.3
attrs                              25.4.0
av                                 16.0.1
bddl                               3.6.0
boto                               2.49.0
boto3                              1.40.69
botocore                           1.40.69
cachetools                         5.5.2
certifi                            2025.10.5
cffi                               2.0.0
cfgv                               3.4.0
charset-normalizer                 3.4.4
click                              8.3.0
cloudpickle                        3.1.2
colorful                           0.6.0a1
contourpy                          1.3.3
crcmod                             1.7
cryptography                       43.0.3
cycler                             0.12.1
dacite                             1.9.2
datasets                           4.4.1
decorator                          5.2.1
deprecated                         1.3.1
diffusers                          0.30.3
dill                               0.4.0
distlib                            0.4.0
dlimp                              0.0.1
dm-tree                            0.1.9
docstring-parser                   0.17.0
draccus                            0.8.0
easydict                           1.13
einops                             0.8.1
etils                              1.13.0
evdev                              1.9.2
executing                          2.2.1
farama-notifications               0.0.4
fast-kinematics                    0.2.2
fastapi                            0.121.0
fasteners                          0.20
fastjsonschema                     2.21.2
filelock                           3.20.0
flash-attn                         2.7.4.post1
flatbuffers                        25.9.23
fonttools                          4.60.1
frozenlist                         1.8.0
fsspec                             2025.10.0
future                             1.0.0
gast                               0.6.0
gcs-oauth2-boto-plugin             3.3
gitdb                              4.0.12
gitpython                          3.1.45
glfw                               2.10.0
google-api-core                    2.28.1
google-apitools                    0.5.35
google-auth                        2.39.0
google-auth-httplib2               0.2.1
google-auth-oauthlib               1.2.3
google-pasta                       0.2.0
google-reauth                      0.1.1
googleapis-common-protos           1.72.0
grpcio                             1.76.0
gsutil                             5.35
gym                                0.26.2
gym-notices                        0.1.0
gymnasium                          0.29.1
h11                                0.16.0
h5py                               3.15.1
hf-xet                             1.2.0
httpcore                           1.0.9
httplib2                           0.20.4
httpx                              0.28.1
huggingface-hub                    0.36.0
hydra-core                         1.4.0.dev1
identify                           2.6.15
idna                               3.11
imageio                            2.37.2
imageio-ffmpeg                     0.6.0
immutabledict                      4.2.2
importlib-metadata                 8.0.0
importlib-resources                6.5.2
iniconfig                          2.3.0
ipython                            9.7.0
ipython-pygments-lexers            1.1.1
jedi                               0.19.2
jinja2                             3.1.6
jmespath                           1.0.1
joblib                             1.5.2
json-numpy                         2.1.1
jsonlines                          4.0.0
jsonschema                         4.25.1
jsonschema-specifications          2025.9.1
jupyter-core                       5.9.1
jupytext                           1.18.1
keras                              2.15.0
kiwisolver                         1.4.10rc0
latex2sympy2                       1.9.0
libclang                           18.1.1
libero                             0.1.0
liger-kernel                       0.6.3
llvmlite                           0.44.0
lxml                               6.0.2
mani-skill                         3.0.0b21
markdown                           3.10
markdown-it-py                     4.0.0
markupsafe                         3.0.3
matplotlib                         3.10.7
matplotlib-inline                  0.2.1
mdit-py-plugins                    0.5.0
mdurl                              0.1.2
mergedeep                          1.3.4
ml-dtypes                          0.2.0
monotonic                          1.6
mplib                              0.1.1
mpmath                             1.3.0
msgpack                            1.1.2
mujoco                             3.3.7
multidict                          6.7.0
multiprocess                       0.70.18
mypy-extensions                    1.1.0
namex                              0.1.0
nbformat                           5.10.4
networkx                           3.6rc0
ninja                              1.13.0
nltk                               3.9.2
nodeenv                            1.9.1
numba                              0.61.2
numpy                              1.26.4
nvidia-cublas-cu12                 12.4.5.8
nvidia-cuda-cupti-cu12             12.4.127
nvidia-cuda-nvrtc-cu12             12.4.127
nvidia-cuda-runtime-cu12           12.4.127
nvidia-cudnn-cu12                  9.1.0.70
nvidia-cufft-cu12                  11.2.1.3
nvidia-curand-cu12                 10.3.5.147
nvidia-cusolver-cu12               11.6.1.9
nvidia-cusparse-cu12               12.3.1.170
nvidia-cusparselt-cu12             0.6.2
nvidia-ml-py                       13.580.82
nvidia-nccl-cu12                   2.21.5
nvidia-nvjitlink-cu12              12.4.127
nvidia-nvtx-cu12                   12.4.127
nvitop                             1.5.3
oauth2client                       4.1.3
oauthlib                           3.3.1
omegaconf                          2.4.0.dev4
opencensus                         0.11.4
opencensus-context                 0.2.dev0
opencv-python                      4.12.0.88
openexr                            3.4.3
opentelemetry-api                  1.26.0
opentelemetry-exporter-prometheus  0.47b0
opentelemetry-proto                1.38.0
opentelemetry-sdk                  1.26.0
opentelemetry-semantic-conventions 0.47b0
openvla-oft                        0.0.1
opt-einsum                         3.4.0
optree                             0.17.0
orjson                             3.11.4
packaging                          25.0
pandas                             2.3.3
parso                              0.8.5
peft                               0.11.1
pexpect                            4.9.0
pillow                             12.0.0
pip                                25.3
platformdirs                       4.5.0
pluggy                             1.6.0
pre-commit                         4.3.0
prettytable                        3.16.0
prometheus-client                  0.23.1
promise                            2.3
prompt-toolkit                     3.0.52
propcache                          0.4.1
proto-plus                         1.26.1
protobuf                           4.25.8
psutil                             7.1.3
ptyprocess                         0.7.0
pure-eval                          0.2.3
py-spy                             0.4.1
pyarrow                            22.0.0
pyasn1                             0.6.1
pyasn1-modules                     0.4.2
pybind11                           3.0.1
pycparser                          2.23
pydantic                           2.12.4
pydantic-core                      2.41.5
pyecharts                          2.0.9
pygments                           2.19.2
pylatexenc                         3.0a33
pynput                             1.8.1
pyopengl                           3.1.10
pyopenssl                          24.2.1
pyparsing                          3.3.0a1
pyperclip                          1.11.0
pytest                             8.4.2
python-dateutil                    2.9.0.post0
python-xlib                        0.33
pytorch-kinematics                 0.7.6
pytorch-seed                       0.2.0
pytz                               2025.2
pyu2f                              0.1.5
pyvers                             0.1.0
pyyaml                             6.0.3
pyyaml-include                     1.4.1
ray                                2.47.1
referencing                        0.37.0
regex                              2025.11.3
requests                           2.32.5
requests-oauthlib                  2.0.0
retry-decorator                    2.0a1
rich                               13.9.4
robosuite                          1.4.1
rpds-py                            0.28.0
rsa                                4.7.2
ruff                               0.14.4
s3transfer                         0.14.0
safetensors                        0.7.0rc0
sapien                             3.0.1
scipy                              1.16.3
sentencepiece                      0.1.99
sentry-sdk                         3.0.0a7
setuptools                         75.8.2
simple-parsing                     0.1.7
simplejson                         3.20.2
six                                1.17.0
smart-open                         7.4.4
smmap                              5.0.2
sniffio                            1.3.1
stack-data                         0.6.3
starlette                          0.49.3
swanlab                            0.7.1
sympy                              1.13.1
tabulate                           0.9.0
tensorboard                        2.15.2
tensorboard-data-server            0.7.2
tensordict                         0.10.0
tensorflow                         2.15.0
tensorflow-addons                  0.23.0
tensorflow-datasets                4.9.3
tensorflow-estimator               2.15.0
tensorflow-graphics                2021.12.3
tensorflow-io-gcs-filesystem       0.37.1
tensorflow-metadata                1.17.2
termcolor                          3.2.0
timm                               0.9.10
tokenizers                         0.19.1
toml                               0.10.2
toppra                             0.6.3
torch                              2.6.0
torch-memory-saver                 0.0.9
torchaudio                         2.6.0
torchdata                          0.11.0
torchvision                        0.21.0
tqdm                               4.67.1
traitlets                          5.14.3
transformers                       4.40.1
transforms3d                       0.4.2
trimesh                            4.9.0
triton                             3.2.0
typeguard                          2.13.3
typing-extensions                  4.15.0
typing-inspect                     0.9.0
typing-inspection                  0.4.2
tyro                               0.10.0a4
tzdata                             2025.2
urllib3                            2.5.0
uvicorn                            0.38.0
virtualenv                         20.35.4
wandb                              0.22.3
wcwidth                            0.2.14
werkzeug                           3.1.3
wheel                              0.45.1
word2number                        1.1
wrapt                              1.14.2
xxhash                             3.6.0
yarl                               1.22.0
zipp                               3.23.0
nvidia-smi:
Wed Dec  3 21:24:16 2025       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.171.04             Driver Version: 535.171.04   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
RLinf version:
v0.1
Docker image tag: 
rlinf/rlinf:agentic-rlinf0.1-torch2.6.0-openvla-openvlaoft-pi0

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).
